core/src/main/java/com/datastrato/gravitino/authorization/PermissionManager.java

```diff
@@ -0,0 +1,305 @@
+            List<RoleEntity> roleEntities =
+                AuthorizationUtils.getValidRoles(
+                    metalake, groupEntity.roles(), groupEntity.roleIds());
+            List<String> roleNames = Lists.newArrayList(toRoleNames(roleEntities));
+            List<Long> roleIds = Lists.newArrayList(toRoleIds(roleEntities));
+
+            if (roleNames.contains(roleEntity.name())) {
+              throw new RoleAlreadyExistsException(
+                  "Role %s already exists in the group %s of the metalake %s",
+                  role, group, metalake);
+            }
+
+            AuditInfo auditInfo =
+                AuditInfo.builder()
+                    .withCreator(groupEntity.auditInfo().creator())
+                    .withCreateTime(groupEntity.auditInfo().createTime())
+                    .withLastModifier(PrincipalUtils.getCurrentPrincipal().getName())
+                    .withLastModifiedTime(Instant.now())
+                    .build();
+
+            roleNames.add(roleEntity.name());
+            roleIds.add(roleEntity.id());
+            return GroupEntity.builder()
+                .withId(groupEntity.id())
+                .withNamespace(groupEntity.namespace())
+                .withName(groupEntity.name())
+                .withRoleNames(roleNames)
+                .withRoleIds(roleIds)
+                .withAuditInfo(auditInfo)
+                .build();
+          });
+      return true;
+    } catch (NoSuchEntityException nse) {
+      LOG.warn("Fail to revoke, group {} does not exist in the metalake {}", group, metalake, nse);
+      throw new NoSuchGroupException(GROUP_DOES_NOT_EXIST_MSG, group, metalake);
+    } catch (IOException ioe) {
+      LOG.error(
+          "Revoking role {} from  group {} failed in the metalake {} due to storage issues",
+          role,
+          group,
+          metalake,
+          ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  public boolean revokeRoleFromGroup(String metalake, String role, String group) {
+    try {
+      AuthorizationUtils.checkMetalakeExists(metalake);
+
+      checkRoleExists(metalake, role);
+
+      RoleEntity roleEntity =
+          store.get(
+              NameIdentifierUtils.ofRole(metalake, role), Entity.EntityType.ROLE, RoleEntity.class);
+
+      AtomicBoolean removed = new AtomicBoolean(true);
+
+      store.update(
+          NameIdentifierUtils.ofGroup(metalake, group),
+          GroupEntity.class,
+          Entity.EntityType.GROUP,
+          groupEntity -> {
+            List<RoleEntity> roleEntities =
+                AuthorizationUtils.getValidRoles(
+                    metalake, groupEntity.roles(), groupEntity.roleIds());
+            List<String> roleNames = Lists.newArrayList(toRoleNames(roleEntities));
+            List<Long> roleIds = Lists.newArrayList(toRoleIds(roleEntities));
+            roleNames.remove(roleEntity.name());
+            removed.set(roleIds.remove(roleEntity.id()));
```

Comment:
What's purpose of making this atomic?

Reply:
This variable is used in the lamda function. This must be final.